packages/core/graph/src/AdjacencyList.js

```diff
@@ -0,0 +1,1195 @@
+
+  linkEdge(
+    prev: EdgeHash | EdgeIndex,
+    edge: EdgeIndex,
+    direction?:
+      | typeof NEXT_HASH
+      | typeof NEXT_IN
+      | typeof PREV_IN
+      | typeof NEXT_OUT
+      | typeof PREV_OUT,
+  ): void {
+    if (direction) {
+      this.#edges[prev + direction] = edge;
+    } else {
+      this.#edges[hashToIndex(prev)] = edge;
+    }
+  }
+
+  unlinkEdge(
+    prev: EdgeHash | EdgeIndex,
+    direction?:
+      | typeof NEXT_HASH
+      | typeof NEXT_IN
+      | typeof PREV_IN
+      | typeof NEXT_OUT
+      | typeof PREV_OUT,
+  ): void {
+    if (direction) {
+      this.#edges[prev + direction] = 0;
+    } else {
+      this.#edges[hashToIndex(prev)] = 0;
+    }
+  }
+
+  /** Get the edge this `edge` links to in the given direction. */
+  getLinkedEdge(
+    prev: EdgeHash | EdgeIndex | null,
+    direction?:
+      | typeof NEXT_HASH
+      | typeof NEXT_IN
+      | typeof PREV_IN
+      | typeof NEXT_OUT
+      | typeof PREV_OUT,
+  ): EdgeIndex | null {
+    if (prev === null) return null;
+    if (direction) {
+      return this.#edges[prev + direction] || null;
+    } else {
+      return this.#edges[hashToIndex(prev)] || null;
+    }
+  }
+
+  /** Find the edge linked to the given `edge`. */
+  findEdgeBefore(
+    edge: EdgeIndex,
+    direction: typeof NEXT_HASH,
+  ): EdgeIndex | null {
+    let candidate = this.getLinkedEdge(this.getHash(edge));
+    if (edge === candidate) return null;
+    while (candidate) {
+      let next = this.getLinkedEdge(candidate, direction);
+      if (next === edge) return candidate;
+      candidate = next;
+    }
+    return null;
+  }
+
+  /**
+   * Adds a node to the graph.
+   *
+   * Returns the id of the added node.
+   */
+  addNode(): NodeId {
+    let id = this.#nodes[COUNT];
+    this.#nodes[COUNT]++;
+    // If we're in danger of overflowing the `nodes` array, resize it.
+    if (this.#nodes[COUNT] >= this.#nodes[CAPACITY]) {
+      // The size of `nodes` doubles every time we reach the current capacity.
+      // This means in the worst case, we will have `O(n - 1)` _extra_
+      // space allocated where `n` is a number nodes that is 1 more
+      // than the previous capacity.
+      this.resizeNodes(increaseNodeCapacity(this.#nodes[CAPACITY]));
+    }
+    return toNodeId(id);
+  }
+
+  /**
+   * Adds an edge to the graph.
+   *
+   * Returns `true` if the edge was added,
+   * or `false` if the edge already exists.
+   */
+  addEdge(
+    from: NodeId,
+    to: NodeId,
+    type: TEdgeType | NullEdgeType = 1,
+  ): boolean {
+    if (fromNodeId(from) < 0 || fromNodeId(from) >= this.#nodes[COUNT]) {
+      throw new Error(`Unknown node ${String(from)}`);
+    }
+    if (fromNodeId(to) < 0 || fromNodeId(to) >= this.#nodes[COUNT]) {
+      throw new Error(`Unknown node ${String(to)}`);
+    }
+    if (type <= 0) throw new Error(`Unsupported edge type ${0}`);
+
+    // The edge is already in the graph; do nothing.
+    if (this.hasEdge(from, to, type)) return false;
```

Comment:
I think this results in hashing twice when the edge is not already in the graph (the common case). You could inline this check below after hashing maybe?